pkg/scheduler/framework/v1alpha1/interface.go

```diff
@@ -273,6 +280,20 @@ type FilterPlugin interface {
 	Filter(ctx context.Context, state *CycleState, pod *v1.Pod, nodeInfo *NodeInfo) *Status
 }
 
+// PostFilterPlugin is an interface for PostFilter plugins. These plugins are called
+// after a pod cannot be scheduled.
+type PostFilterPlugin interface {
+	Plugin
+	// PostFilter is called by the scheduling framework.
+	// A PostFilter should return one of the following statuses:
+	// - Unschedulable: a plugin ran its logic but still cannot make it schedulable.
+	// - Schedulable: a plugin ran its logic and can make it schedulable.
+	// - Error: some internal error occurs.
+	// - Success: indicate the plugin ran its logic successfully, but it has no idea whether
+	//   the pod is schedulable or not.
```

Comment:
> I think a PostFilter plugin should return either Success or Failure statuses only.

So you're implying `Failure` == internal error, and `Success` to be categorized by the first parameter (nominatedNodename or PostFilterResult):
- success and continue: current plugin cannot make the pod schedulable
- success and return: current plugin can make the pod schedulable

> The other parameter should probably be a generic PostFilterResult with entries for cases that we support natively, the first of which is PreemptionResult.

I agree PostFilterResult is generic, but I feel like it's sort of premature optimization - NominatedNodeName has to be retrieved from PostFilterResult -> PreemptionResult -> NominatedNodeName. I think we should start with the minimum (a string) data struct, and involve it only when necessary.

Reply:
> success and continue: current plugin cannot make the pod schedulable
> success and return: current plugin can make the pod schedulable

I am suggesting to do the logic in the first comment. RunPostFilterPlugins should execute all plugins unless one returns an Error (internal failure). Results get overriden if multiple ones were executed and successful. The idea is that PostFilter is generic, not all PostFilter plugins are Preemption plugins.

> I agree PostFilterResult is generic, but I feel like it's sort of premature optimization - NominatedNodeName has to be retrieved from PostFilterResult -> PreemptionResult -> NominatedNodeName. I think we should start with the minimum (a string) data struct, and involve it only when necessary.

This is an interface not an optimization though, we need to get it properly designed from the beginning. If we agree that PostFilter is a generic extension point, not one that only executes Preemption, then we should try to design the interface for that, otherwise we should just call the extension point preemption.

Reply:
> RunPostFilterPlugins should execute all plugins unless one returns an Error (internal failure).

Users may configure more than one PostFilter plugins to make the pod schedulable. For example, the first one may be reusing the upstream defaultpreemption which tries strategies on a single node; and the second one may be a custom one o try complex (and hence expensive) preemption logic; the third one may be trying some CA/VPA strategies. So instead of running all plugins, it's good to stop at the first success as we've got one available solution. So a binary return value Success or Failure cannot expose enough info. An additional status Schedulable/Unschedulable would help. Or, it's possible to _infer_ it by checking the preemptionResult. But that's not generic as some strategy (like CA/VPA) is not related with preemption. 

> If we agree that PostFilter is a generic extension point, not one that only executes Preemption, then we should try to design the interface for that, otherwise we should just call the extension point preemption.

We cannot go back, right? (kidding...) Generally, I'm ok with this, one thing is that I think we can design `PostFilterResult` as an interface, and now it has only one function `NominatedNodeName()`. In this way, we don't force the implementation to be using a "preemptionResult" as some plugin are not literally related with "preemption" - such as CA (add a new machine) or VPA (to reduce cpu/mem requests of some Pods).

Reply:
ok, my thinking was that we want PostFilter to be exactly like the extension point we used to have, but now that we have PreScore, I guess we can have its semantics more strict to address the cases you mentioned: updating the cluster state to potentially make the pod schedulable.

In this case, I agree that PostFilter should run only when Filter return no nodes, which means only one of PostFilter or PreScore is executed in a given scheduling cycle. The name PostFilter is a little misleading though if we agree that this is the semantics we want.

Reply:
> In this case, I agree that PostFilter should run only when Filter return no nodes

Yup, in the first version I just assumed PostFilter to be executed when Filter failed.

> which means only one of PostFilter or PreScore is executed in a given scheduling cycle

Yes, they're mutually exclusive.

> The name PostFilter is a little misleading

`ResolveFilterError` may be more descriptive. But I think `PostFiler` is ok with proper documenting.

Reply:
I think ```Schedulable``` and ```Success``` have overlapping semantics, so I don't think we should introduce a new status. We can support the following statuses for PostFilter:

1) ```Unschedulable```: continue to execute other plugins
2) ```Success```: pod is potentially schedulable, rest of plugins are not executed.
3) ```Error```: internal error, rest of plugins are not executed.

So, the signatures and RunPostFilterPlugins could look like the following:

```
type  PostFilterResult struct {
  NominatedNodeName string   // if set, preemption found a candidate node.
}

PostFilter(ctx context.Context, state *CycleState, pod *v1.Pod, ...) (*PostFilterResult, *Status)

```

and RunPostFilterPlugins would look like this:
```
func (f *framework) RunPostFilterPlugins(...) *PostFilterResult, *Status {
   statuses := make(PluginToStatus) 
   for _, pl := range f.postFilterPlugins {
         r, s := pl(...); s.IsSuccess()
         if  s.IsSuccess() {
              return r, s
         } else if !s.IsUnschedulable() {
             return nil, NewStatus(Error, s.Message())
         }
         statuses[p.Name()] = s
   }
   return nil, statuses.Merge()
}
```

Reply:
> Unschedulable: continue to execute other plugins

We should document that "informational" plugins should be placed to run first and return `Unschedulable` if they succeed.

Reply:
@ahg-g the 3-status set looks good to me.

@alculquicondor true.